server/internal/worker/client.go

```diff
@@ -0,0 +1,383 @@
+// One approach is to have each ClientHandle have a global marker of some kind
+// that both the getter and the submitter check once per loop to see if they
+// should return. However, this mechanism requires that network reads and writes
+// time out quickly so that the loop is executed frequently. This in turn
+// requires intermediate partial reads or writes to be buffered and then
+// completed on the next loop. For writing, this isn't so bad - it's just a
+// matter of keeping track of how many bytes were written already. For reading,
+// it's somewhat more complicated, requiring keeping partial parse state.
+//
+// Instead, we sidestep this issue by allowing network reads and writes to block
+// as long as they want to, and instead using the Destroy method, which closes
+// the network connection. Closing a network connection causes all existing
+// pending reads and writes to immediately return, so this allows us to simply
+// call Destroy and know that outstanding calls to GetJob or SubmitResult will
+// immediately return.
+
+// A Client is a server-side handle on a connection to a client. It is capable
+// of reading job requests from the client, writing responses to the client, and
+// shutting down the connection.
+//
+// Since Clients can be backed by TCP connections, it is acceptable for GetJob
+// to return false (indicating that there are no jobs to be done - in the case
+// of a TCP connection, caused by a half-close of the connection) while
+// SubmitResult continues to return true (because writing to the connection is
+// still allowed).
+type Client interface {
+	// GetJob returns the next job to perform, or ok = false if there are no more
+	// jobs to do. If it returns a job, it also returns the pool to which that job
+	// should be submitted.
+	GetJob() (job interface{}, pool *Pool, ok bool)
+
+	// SubmitResult submits the result of a previously-requested job. If the
+	// result cannot be submitted, it returns false. This implies that the client
+	// is completely closed and no results can ever be submitted in the future.
+	SubmitResult(result interface{}) (ok bool)
+
+	// IsAlive returns whether or not the client is still alive (if no more
+	// jobs can be gotten via GetJob, but submitting is still enabled, the client
+	// is considered alive). Once it returns false, it may never return true
+	// again.
+	//
+	// It must be safe to call IsAlive so long as the Client object exists.
+	IsAlive() bool
+
+	// Destroy destructs the client's connection. After a call to Destroy, all
+	// pending calls to GetJob and SubmitResult should immediately return false,
+	// and any future calls to those methods should also immediately return false.
+	//
+	// It must be safe to call Destroy multiple times; all calls after the first
+	// are no-ops.
+	Destroy()
+}
+
+// A ClientHandle is a handle on a pair of reader/writer goroutines that are
+// processing requests from a client.
+type ClientHandle struct {
```

Comment:
This is unfortunately much more complicated than that. `net.Conn` allows you to perform multiple operations on a connection at a time. What we're doing here is abstracting the concept of receiving requests for work from a client, performing those requests, and sending back responses. The functionality here also includes out-of-band backpressure to avoid DoS vulnerabilities (see the long comment at the top of the file).